### ISSUE TITLE
Predefined sizes for padding, width, and height

### ISSUE THREAD

**Issue by kawaijoe** [duobaseio/forui#91]
### Description
Including some constants that can be used in place of magic values would be nice. These constant values can be used for widgets' padding, width, and height.

In web development, Tailwind is commonly used and it has a decent solution. Tailwind example usage:

- `w-0` for a width of 0px
- `w-0.5` for a width of 2px
- `w-1` for a width of 4px
- `w-1.5` for a width of 6px
- `w-2`, `w-2.5`, `w-3`, ...

Ideally, it would be great to adopt something similar to what Tailwind is doing. However, we noticed a major issue when trying to mimick a similar model. Tailwind uses a string to represent these variations and has no issues with using a dot separator to represent half (eg. `0.5`).

On the other hand, a dot in Dart represents an accessor. In addition, variables in Dart cannot start with a number.

Any input or suggestions are greatly appreciated!

### Possible solutions

- **Bump values**
eg. bump `0.5` values -> `1` and `1` -> `2`.
While this will generally work, it breaks down with the generic heuristic used. Generally, `1` represents `4px` in Tailwind, and `0.5` is exclusively used for smaller values to represent `2px`.
Another great alternative we're leaning towards is to make `1` represent `2px`. However, that may deviate from what Tailwind users expect.

- **Suffix characters to represent half**
eg. bump `0.5` values -> `0h` OR `0.5` values -> `0Half`.
Might not look/feel the best.

### Reference links

- https://tailwindcss.com/docs/width
- https://tailwindcss.com/docs/height
- https://tailwindcss.com/docs/padding

**Comment by JoDeveloper** [comment]
something like this ?

```dart
import 'package:flutter/material.dart';

class FSizes {
  // Widths and Heights
  static const w0 = 0.0;
  static const w1 = 4.0;
  static const w2 = 8.0;
  static const w3 = 12.0;
  static const w4 = 16.0;
  static const w5 = 20.0;
  static const w6 = 24.0;
  static const w8 = 32.0;
  static const w10 = 40.0;
  static const w12 = 48.0;
  static const w16 = 64.0;
  static const w20 = 80.0;
  static const w24 = 96.0;
  static const w32 = 128.0;
  static const w40 = 160.0;
  static const w48 = 192.0;
  static const w56 = 224.0;
  static const w64 = 256.0;
  static const w72 = 288.0;
  static const w80 = 320.0;
  static const w96 = 384.0;

  // Heights (same as widths for consistency)
  static const h0 = w0;
  static const h1 = w1;
  static const h2 = w2;
  static const h3 = w3;
  static const h4 = w4;
  static const h5 = w5;
  static const h6 = w6;
  static const h8 = w8;
  static const h10 = w10;
  static const h12 = w12;
  static const h16 = w16;
  static const h20 = w20;
  static const h24 = w24;
  static const h32 = w32;
  static const h40 = w40;
  static const h48 = w48;
  static const h56 = w56;
  static const h64 = w64;
  static const h72 = w72;
  static const h80 = w80;
  static const h96 = w96;

  // Paddings
  static const p0 = 0.0;
  static const p1 = 4.0;
  static const p2 = 8.0;
  static const p3 = 12.0;
  static const p4 = 16.0;
  static const p5 = 20.0;
  static const p6 = 24.0;
  static const p8 = 32.0;
  static const p10 = 40.0;
  static const p12 = 48.0;
  static const p16 = 64.0;
  static const p20 = 80.0;
  static const p24 = 96.0;
  static const p32 = 128.0;
  static const p40 = 160.0;
  static const p48 = 192.0;
  static const p56 = 224.0;
  static const p64 = 256.0;

  // Margin (same as padding for consistency)
  static const m0 = p0;
  static const m1 = p1;
  static const m2 = p2;
  static const m3 = p3;
  static const m4 = p4;
  static const m5 = p5;
  static const m6 = p6;
  static const m8 = p8;
  static const m10 = p10;
  static const m12 = p12;
  static const m16 = p16;
  static const m20 = p20;
  static const m24 = p24;
  static const m32 = p32;
  static const m40 = p40;
  static const m48 = p48;
  static const m56 = p56;
  static const m64 = p64;

  // Percentage-based sizes
  static double wp(BuildContext context, double percent) =>
      MediaQuery.of(context).size.width * percent / 100;

  static double hp(BuildContext context, double percent) =>
      MediaQuery.of(context).size.height * percent / 100;

  // Padding helpers
  static EdgeInsets pa(double value) => EdgeInsets.all(value);
  static EdgeInsets ph(double value) => EdgeInsets.symmetric(horizontal: value);
  static EdgeInsets pv(double value) => EdgeInsets.symmetric(vertical: value);
  static EdgeInsets pt(double value) => EdgeInsets.only(top: value);
  static EdgeInsets pr(double value) => EdgeInsets.only(right: value);
  static EdgeInsets pb(double value) => EdgeInsets.only(bottom: value);
  static EdgeInsets pl(double value) => EdgeInsets.only(left: value);

  // Margin helpers
  static EdgeInsets ma(double value) => EdgeInsets.all(value);
  static EdgeInsets mh(double value) => EdgeInsets.symmetric(horizontal: value);
  static EdgeInsets mv(double value) => EdgeInsets.symmetric(vertical: value);
  static EdgeInsets mt(double value) => EdgeInsets.only(top: value);
  static EdgeInsets mr(double value) => EdgeInsets.only(right: value);
  static EdgeInsets mb(double value) => EdgeInsets.only(bottom: value);
  static EdgeInsets ml(double value) => EdgeInsets.only(left: value);

}
```

**Comment by kawaijoe** [comment]
@JoDeveloper thanks for the suggestion.

I think the values need not be repeated for padding/margin because they represent the same thing. However, it doesn't quite solve the `.5` issue :(

I'm personally not too big of a fan of:
```dart
static EdgeInsets pa(double value) => EdgeInsets.all(value);
static EdgeInsets ph(double value) => EdgeInsets.symmetric(horizontal: value);
static EdgeInsets pv(double value) => EdgeInsets.symmetric(vertical: value);
static EdgeInsets pt(double value) => EdgeInsets.only(top: value);
...
```

While it does make the code shorter, I think it potentially makes the codebase harder to understand compared to using the `EdgeInsets` constructor directly.

**Comment by JoDeveloper** [comment]
https://github.com/forus-labs/forui/issues/91#issuecomment-2241141839

```dart
// Base unit (2px)
  static const double _unit = 2.0;

  // Width constants
  static const double w0 = 0;
  static const double w05 = _unit * 0.5;  // 1px
  static const double w1 = _unit;         // 2px
  static const double w15 = _unit * 1.5;  // 3px
  static const double w2 = _unit * 2;     // 4px
  static const double w25 = _unit * 2.5;  // 5px
  static const double w3 = _unit * 3;     // 6px
  static const double w35 = _unit * 3.5;  // 7px
  static const double w4 = _unit * 4;     // 8px
  static const double w5 = _unit * 5;     // 10px
  static const double w6 = _unit * 6;     // 12px
  static const double w7 = _unit * 7;     // 14px
  static const double w8 = _unit * 8;     // 16px
  static const double w9 = _unit * 9;     // 18px
  static const double w10 = _unit * 10;   // 20px
  static const double w11 = _unit * 11;   // 22px
  static const double w12 = _unit * 12;   // 24px
  static const double w14 = _unit * 14;   // 28px
  static const double w16 = _unit * 16;   // 32px
  static const double w20 = _unit * 20;   // 40px
  static const double w24 = _unit * 24;   // 48px
  static const double w28 = _unit * 28;   // 56px
  static const double w32 = _unit * 32;   // 64px
  static const double w36 = _unit * 36;   // 72px
  static const double w40 = _unit * 40;   // 80px
  static const double w44 = _unit * 44;   // 88px
  static const double w48 = _unit * 48;   // 96px
  static const double w52 = _unit * 52;   // 104px
  static const double w56 = _unit * 56;   // 112px
  static const double w60 = _unit * 60;   // 120px
  static const double w64 = _unit * 64;   // 128px
  static const double w72 = _unit * 72;   // 144px
  static const double w80 = _unit * 80;   // 160px
  static const double w96 = _unit * 96;   // 192px
  
  ```

**Comment by kawaijoe** [comment]
> [#91 (comment)](https://github.com/forus-labs/forui/issues/91#issuecomment-2241141839)
> 
> ```dart
> // Base unit (2px)
>   static const double _unit = 2.0;
> 
>   // Width constants
>   static const double w0 = 0;
>   static const double w05 = _unit * 0.5;  // 1px
>   static const double w1 = _unit;         // 2px
>   static const double w15 = _unit * 1.5;  // 3px
>   static const double w2 = _unit * 2;     // 4px
>   static const double w25 = _unit * 2.5;  // 5px
>   static const double w3 = _unit * 3;     // 6px
>   static const double w35 = _unit * 3.5;  // 7px
>   static const double w4 = _unit * 4;     // 8px
>   static const double w5 = _unit * 5;     // 10px
>   static const double w6 = _unit * 6;     // 12px
>   static const double w7 = _unit * 7;     // 14px
>   static const double w8 = _unit * 8;     // 16px
>   static const double w9 = _unit * 9;     // 18px
>   static const double w10 = _unit * 10;   // 20px
>   static const double w11 = _unit * 11;   // 22px
>   static const double w12 = _unit * 12;   // 24px
>   static const double w14 = _unit * 14;   // 28px
>   static const double w16 = _unit * 16;   // 32px
>   static const double w20 = _unit * 20;   // 40px
>   static const double w24 = _unit * 24;   // 48px
>   static const double w28 = _unit * 28;   // 56px
>   static const double w32 = _unit * 32;   // 64px
>   static const double w36 = _unit * 36;   // 72px
>   static const double w40 = _unit * 40;   // 80px
>   static const double w44 = _unit * 44;   // 88px
>   static const double w48 = _unit * 48;   // 96px
>   static const double w52 = _unit * 52;   // 104px
>   static const double w56 = _unit * 56;   // 112px
>   static const double w60 = _unit * 60;   // 120px
>   static const double w64 = _unit * 64;   // 128px
>   static const double w72 = _unit * 72;   // 144px
>   static const double w80 = _unit * 80;   // 160px
>   static const double w96 = _unit * 96;   // 192px
>   
> ```

I don't think that's particularly great as well. For example, `w15` may be interpreted as `15` instead of `1.5`.

**Comment by JoDeveloper** [comment]
> I don't think that's particularly great as well. For example, `w15` may be interpreted as `15` instead of `1.5`.

other suggestions maybe :

```
static const double w1 = _unit;         // 1px
static const double w1Half = _unit * 1.5; // 1.5px

or :

static const double w1x = _unit;       // 1px
static const double w1y = _unit * 1.5; // 1.5px

what do you think 🤔 ! 
```

**Comment by kawaijoe** [comment]
I'm unsure if the `x` and `y` are good representations. `Half` was one of the original suggestions!

**Comment by a-man-called-q** [comment]
Hi, 

I'm interested with this idea, and here's my cents :

- Tailwind's w / h ( like w-0 or w-full ) is basically units that came out from grid based design in order to help responsive first web design.
- A single web app / page, by default is a combination of Javascript, CSS, html. 3 different language with 3 completely different syntax and ways of implementing.
- Flutter can't support XML based syntax by default  ( Although there's [this](https://pub.dev/packages/xml_widget) ) but it's practically only support Classes and Functions. Imagine React but without JSX lol.

My take on this is along the line of what @JoDeveloper suggested . Something like

```dart
class FSize {
  static double _defaultUnit = 2.0;

  static void setDefaultGrid(double unit) => _defaultUnit = unit;

  static double _resolveGrid(BuildContext? context) {
    if (context == null) return _defaultUnit;
    try {
      return FTheme.of(context).spacingUnit;
    } catch (e) {
      assert(() {
        debugPrint('[FSize] ⚠️ No FTheme found above context. Using default grid ($_defaultUnit).');
        return true;
      }());
      return _defaultUnit;
    }
  }

  static double w({required double units, BuildContext? context}) =>
      _resolveGrid(context) * units;

  static double h({required double units, BuildContext? context}) =>
      _resolveGrid(context) * units;
}
```

and for the spacing
```dart
class FSpace {
  static EdgeInsets p({required double units, BuildContext? context}) =>
      EdgeInsets.all(FSize.w(units: units, context: context));

  static EdgeInsets ph({required double units, BuildContext? context}) =>
      EdgeInsets.symmetric(horizontal: FSize.w(units: units, context: context));

  static EdgeInsets pv({required double units, BuildContext? context}) =>
      EdgeInsets.symmetric(vertical: FSize.h(units: units, context: context));

  static EdgeInsets pt({required double units, BuildContext? context}) =>
      EdgeInsets.only(top: FSize.h(units: units, context: context));

  static EdgeInsets pr({required double units, BuildContext? context}) =>
      EdgeInsets.only(right: FSize.w(units: units, context: context));

  static EdgeInsets pb({required double units, BuildContext? context}) =>
      EdgeInsets.only(bottom: FSize.h(units: units, context: context));

  static EdgeInsets pl({required double units, BuildContext? context}) =>
      EdgeInsets.only(left: FSize.w(units: units, context: context));
}
```
using this kind of implementation, you can write something like this 

```dart
FTheme(
  data: FThemes.zinc.light.copyWith(spacingUnit: 6.0), // No need to change the FTheme Class
  child: MaterialApp(
    home: YourPage(),
  ),
);

Container(
  width: FSize.w(units: 5.5, context: context),
  height: FSize.h(units: 2, context: context),
  padding: FSpace.p(units: 1.5, context: context),
  margin: FSpace.ph(units: 2, context: context),
);

//or if you're not changing the default spacingUnit

Container(
  width: FSize.w(units: 5.5),
  height: FSize.h(units: 2),
  padding: FSpace.p(units: 1.5),
  margin: FSpace.ph(units: 2),
);
```

This feels more Flutter-y. However, maybe you might want to make it looks like Tailwind-ish, `FSize.w(units: 5.5, context: context)` might feels too much for many ( including me who do stuff in AstroJS most of the time ), you can implement sugar syntax like

```dart
extension FSizeX on num {
  double get w => FSize.w(units: toDouble());
  double get h => FSize.h(units: toDouble());
}
```
using `setDefaultGrid()` from `FSize` class, we can make something like this

```dart
FSize.setDefaultGrid(4.0);

Container(
  width: 8.w,       // 32px
  height: 3.5.h,    // 14px
  padding: FSpace.p(units: 2),
);
```

why not using static values like ` static const double w05 = _unit * 0.5; `, in my previous experiences.. using tailwind I sometimes encounter cases where i need to use padding or margin or text size that is not written in the static values, and I ended up using something like `p-[30em]`  or being annoyed because there is `p-16` but there's no `p-17`.

**Comment by kawaijoe** [comment]
Hey @a-man-called-q! Thank you for the detailed write up. The initial thought behind creating the static variables was actually to discourage usages of `p-17` and nudge the user towards using `p-16` in hopes that it will produce more consistent user interface.

For example, Mantine UI only has `xs`, `sm`, `md`, `lg`, `xl` which is far more limiting.
https://mantine.dev/styles/style-props/#theme-values

It's worth noting that tailwind is actually moving away from fixed values. In the latest version of tailwind v4, it no longer have the restrictions stated above and you're free to use `p-<number>`.
https://tailwindcss.com/docs/padding

I don't have strong opinions on this at the moment, but the main advantage of using something similar to what @a-man-called-q has proposed would be being able to dynamically change the spacing on the theme level. However, this comes with the added complexity of having an additional class (eg. `FSize`) when the user could have just used the number directly.
``` dart
FThemes.zinc.light.copyWith(spacingUnit: 6.0); // Quoted from above
```

Do you have any thoughts @Pante @Daviiddoo?

**Comment by a-man-called-q** [comment]
> The initial thought behind creating the static variables was actually to discourage usages of `p-17` and nudge the user towards using `p-16` in hopes that it will produce more consistent user interface.
> 
> For example, Mantine UI only has `xs`, `sm`, `md`, `lg`, `xl` which is far more limiting. https://mantine.dev/styles/style-props/#theme-values

Yes I do agree, because consistency is crucial in front end development. However, `It's worth noting that tailwind is actually moving away from fixed values. ` Most probably there are cases that the size can't be limited to `xs` or `lg`.

It also worth mentioning that my solution would have a bit confusing effect for those who used to tailwind. `p-4` means `units * x = 4` while in my proposal `FSize.p(units:4, context)` means `units * 4 = x`.

> I don't have strong opinions on this at the moment, but the main advantage of using something similar to what [@a-man-called-q](https://github.com/a-man-called-q) has proposed would be being able to dynamically change the spacing on the theme level. However, this comes with the added complexity of having an additional class (eg. `FSize`) when the user could have just used the number directly.

`FSize.setDefaultGrid(4.0);` you mean this @kawaijoe ? It's for customizing the default `spacingUnit` ( the default is 2.0 ) if you want to use custom value without `context` and using something like `3.5.h`.

`FThemes.zinc.light.copyWith(spacingUnit: 6.0);` stored the `spacingUnit` inside `context` thus, I don't know how to read the `context` without passing it to the function.

**Comment by Pante** [comment]
Generally, I'm not a huge fan of introducing extensions on core language types. Also I'm just curious but are there any use-cases for wanting to adjust the spacing units?

**Comment by a-man-called-q** [comment]
> Also I'm just curious but are there any use-cases for wanting to adjust the spacing unit?

TBH, it's very rare... but I have met people who used 3 based grid and 4 based grids. However, it is very rare.

If we eliminate the concept of custom spacing unit, then it would reduce the boilerplate. If we combine it with @kawaijoe 's opinion :

```dart
final class FSize {
  static const double defaultUnit = 2.0;

  // Size constants
  static const double s0 = 0;
  static const double s1 = defaultUnit * 1;
  static const double s2 = defaultUnit * 2;
  static const double s3 = defaultUnit * 3;
  static const double s4 = defaultUnit * 4;
  static const double s5 = defaultUnit * 5;
  static const double s6 = defaultUnit * 6;
  static const double s8 = defaultUnit * 8;
  static const double s10 = defaultUnit * 10;
 // ...

  // EdgeInsets constants for common patterns
  static const EdgeInsets edge0 = EdgeInsets.zero;
  static const EdgeInsets edge1 = EdgeInsets.all(s1);
 // ...

  // Horizontal EdgeInsets constants
  static const EdgeInsets edgeX0 = EdgeInsets.symmetric(horizontal: s0);
  static const EdgeInsets edgeX1 = EdgeInsets.symmetric(horizontal: s1);
  static const EdgeInsets edgeX2 = EdgeInsets.symmetric(horizontal: s2);
 // ...

  // Vertical EdgeInsets constants
  static const EdgeInsets edgeY0 = EdgeInsets.symmetric(vertical: s0);
  static const EdgeInsets edgeY1 = EdgeInsets.symmetric(vertical: s1);
 // ...

  /// Calculate a size based on units
  static double s(double units) => defaultUnit * units;

  // EdgeInsets helpers (for both padding and margin)
  
  /// Creates EdgeInsets on all sides
  static EdgeInsets edge(double units) => 
      EdgeInsets.all(units * defaultUnit);
      
  /// Creates horizontal EdgeInsets (left & right)
  static EdgeInsets edgeX(double units) => 
      EdgeInsets.symmetric(horizontal: units * defaultUnit);
      
  /// Creates vertical EdgeInsets (top & bottom)
  static EdgeInsets edgeY(double units) => 
      EdgeInsets.symmetric(vertical: units * defaultUnit);
      
  /// Creates EdgeInsets on top side only
  static EdgeInsets edgeTop(double units) => 
      EdgeInsets.only(top: units * defaultUnit);
      
  /// Creates EdgeInsets on right side only
  static EdgeInsets edgeRight(double units) => 
      EdgeInsets.only(right: units * defaultUnit);
      
  /// Creates EdgeInsets on bottom side only
  static EdgeInsets edgeBottom(double units) => 
      EdgeInsets.only(bottom: units * defaultUnit);
      
  /// Creates EdgeInsets on left side only
  static EdgeInsets edgeLeft(double units) => 
      EdgeInsets.only(left: units * defaultUnit);
      
  /// Creates custom EdgeInsets
  static EdgeInsets edgeOnly({
    double left = 0.0,
    double top = 0.0,
    double right = 0.0,
    double bottom = 0.0,
  }) {
    return EdgeInsets.only(
      left: left * defaultUnit,
      top: top * defaultUnit,
      right: right * defaultUnit,
      bottom: bottom * defaultUnit,
    );
  }
}
```
and we can use something like `FSize.h(1.5)` or `FSize.h1_5` (provided that we add `h1_5` inside the class). and for the edge insets, we can use something like `FSize.edgeX2`

PS : Sorry got flu, so I'm only working on 20% capacity right now.

**Comment by Pante** [comment]
Having spent the past few days making widget sizes consistent on the same platform, I'm unconvinced that size, width and height tokens are useful in the **currently proposed form**. 

I think the biggest motivation behind Tailwind's CSS utilities is allowing components to be styled inline and avoid the complexities revolving CSS styling. This isn't applicable in the context of Flutter/Forui where size, width and height are all controlled via plain Dart code. 

Another argument for spacing tokens is to provide an abstract unit independent of hardware differences like pixel density. This isn't applicable in the context of Flutter either since it already uses logical pixels. Adding a quantized scale on top of logical pixels does not seem useful. 

There are two types of tokens discussed in this issue:
* Granular tokens, e.g. `w-0, w-1, w-2`.
* Coarse semantic tokens, e.g. `xs, sm, md`.

Granular tokens provide fine-grained control over styling. Nevertheless, it provides minimal benefits over hard-coded values. 
`w-4` isn't more readable compared to  `4`. Neither does it convey any semantic information as to why we chose `w-4` over `w-5`. 

An argument for granular tokens is the promise of easily scaling an entire application using a single centralized set of values. However, my time spent on making widget sizes scale consistently across platforms has convinced me that there isn't a single simple linear transformation to achieve this in practice.

Coarse semantic tokens are far too coarse to be useful in practice. There will be several cases where neither the smaller nor larger value is ideal.

TL;DR: High cognitive and developer pain for minimal benefits.
